File: src/Rulesage.Shared/Repositories/Implementations/ActionRepository.cs
using System.Text.Json;
using Microsoft.FSharp.Collections;
using Npgsql;
using NpgsqlTypes;
using Pgvector;
using Rulesage.Common.Grammar.Ast;
using Rulesage.Shared.Repositories.Abstractions;

namespace Rulesage.Shared.Repositories.Implementations;

public class ActionRepository(NpgsqlDataSource dataSource, JsonSerializerOptions jsonOptions) : IActionRepository
{
    public async Task<IEnumerable<string>> GetDocumentsAsync(CancellationToken cancellationToken = default)
    {
        await using var conn = dataSource.CreateConnection();
        await conn.OpenAsync(cancellationToken);

        await using var cmd = new NpgsqlCommand("SELECT annotation FROM actions", conn);
        await using var reader = await cmd.ExecuteReaderAsync(cancellationToken);

        return ReadToList(reader, r => r.GetString(0));
    }

    public async Task<IEnumerable<ActionExpr>> FindByIdsAsync(IEnumerable<string> ids,
        CancellationToken cancellationToken = default)
    {
        await using var conn = dataSource.CreateConnection();
        await conn.OpenAsync(cancellationToken);

        await using var cmd =
            new NpgsqlCommand(
                """
                select
                    id,
                    community_id,
                    annotation,
                    generic_params,
                    fors,
                    returns,
                    script
                from actions
                where id=any($1) 
                """,
                conn
            );

        // ReSharper disable once BitwiseOperatorOnEnumWithoutFlags
        cmd.Parameters.Add(new NpgsqlParameter<string[]>
            { Value = ids.ToArray(), NpgsqlDbType = NpgsqlDbType.Array | NpgsqlDbType.Text });

        await using var reader = await cmd.ExecuteReaderAsync(cancellationToken);

        var idOrdinal = reader.GetOrdinal("id");
        var communityOrdinal = reader.GetOrdinal("community_id");
        var annotationOrdinal = reader.GetOrdinal("annotation");
        var genericParamsOrdinal = reader.GetOrdinal("generic_params");
        var forsOrdinal = reader.GetOrdinal("fors");
        var returnsOrdinal = reader.GetOrdinal("returns");
        var scriptOrdinal = reader.GetOrdinal("script");

        return ReadToList(reader, r =>
        {
            var id = r.GetString(idOrdinal);
            var community = r.GetString(communityOrdinal);
            var annotation = r.GetString(annotationOrdinal);
            var genericParamsJson = r.GetString(genericParamsOrdinal);
            var forsJson = r.GetString(forsOrdinal);
            var returnsJson = r.GetString(returnsOrdinal);
            var script = r.GetString(scriptOrdinal);

            var genericParams =
                JsonSerializer.Deserialize<FSharpList<string>>(genericParamsJson, jsonOptions);
            var fors =
                JsonSerializer.Deserialize<FSharpMap<string, ParamExpr>>(forsJson, jsonOptions);
            var returns =
                JsonSerializer.Deserialize<TypeExpr>(returnsJson, jsonOptions);

            return new ActionExpr(id, community, annotation, genericParams, fors, returns, script);
        });
    }

    public async Task<IEnumerable<(ActionExpr, float)>> FindOrderByCosineDistanceAsync(string contextCommunity,
        float[] queryVector,
        int skip, int take,
        CancellationToken cancellationToken = default)
    {
        await using var conn = dataSource.CreateConnection();
        await conn.OpenAsync(cancellationToken);

        var sections = contextCommunity.Split('.');
        var hierarchy = new string[sections.Length + 1];
        var sectionSum = "";
        hierarchy[0] = "";
        for (var i = 0; i < sections.Length; i++)
        {
            sectionSum += sections[i];
            hierarchy[i + 1] = sectionSum;
        }

        await using var cmd =
            new NpgsqlCommand(
                """
                select
                    id,
                    annotation,
                    generic_params,
                    fors,
                    returns,
                    script,
                    (annotation_embedding <=> $1) as distance
                from actions
                where community_id = any($2)
                order by distance
                limit $3 offset $4;
                """,
                conn
            );

        cmd.Parameters.Add(new NpgsqlParameter { Value = new Vector(queryVector), DataTypeName = "vector" });
        // ReSharper disable once BitwiseOperatorOnEnumWithoutFlags
        cmd.Parameters.Add(new NpgsqlParameter<string[]>
            { Value = hierarchy, NpgsqlDbType = NpgsqlDbType.Array | NpgsqlDbType.Text });
        cmd.Parameters.Add(new NpgsqlParameter<int> { Value = take, NpgsqlDbType = NpgsqlDbType.Integer });
        cmd.Parameters.Add(new NpgsqlParameter<int> { Value = skip, NpgsqlDbType = NpgsqlDbType.Integer });

        await using var reader = await cmd.ExecuteReaderAsync(cancellationToken);

        var idOrdinal = reader.GetOrdinal("id");
        var annotationOrdinal = reader.GetOrdinal("annotation");
        var genericParamsOrdinal = reader.GetOrdinal("generic_params");
        var forsOrdinal = reader.GetOrdinal("fors");
        var returnsOrdinal = reader.GetOrdinal("returns");
        var scriptOrdinal = reader.GetOrdinal("script");
        var distanceOrdinal = reader.GetOrdinal("distance");

        return ReadToList(reader, r =>
        {
            var id = r.GetString(idOrdinal);
            var annotation = r.GetString(annotationOrdinal);
            var genericParamsJson = r.GetString(genericParamsOrdinal);
            var forsJson = r.GetString(forsOrdinal);
            var returnsJson = r.GetString(returnsOrdinal);
            var script = r.GetString(scriptOrdinal);

            var genericParams =
                JsonSerializer.Deserialize<FSharpList<string>>(genericParamsJson, jsonOptions);
            var fors =
                JsonSerializer.Deserialize<FSharpMap<string, ParamExpr>>(forsJson, jsonOptions);
            var returns =
                JsonSerializer.Deserialize<TypeExpr>(returnsJson, jsonOptions);

            return (
                new ActionExpr(id, contextCommunity, annotation, genericParams, fors, returns, script),
                1f - (float)r.GetDouble(distanceOrdinal)
            );
        });
    }

    public async Task<bool> SaveAsync(IEnumerable<ActionExpr> actions, CancellationToken cancellationToken = default)
    {
        await using var conn = dataSource.CreateConnection();
        await conn.OpenAsync(cancellationToken);

        var actionsArray = actions.ToArray();
        var count = actionsArray.Length;

        var ids = new string[count];
        var communities = new string[count];
        var annotations = new string[count];
        var genericParams = new FSharpList<string>[count];
        var fors = new FSharpMap<string, ParamExpr>[count];
        var returns = new TypeExpr[count];
        var scripts = new string[count];

        for (var i = 0; i < count; i++)
        {
            var r = actionsArray[i];
            ids[i] = r.Id;
            communities[i] = r.Community;
            annotations[i] = r.Annotation;
            genericParams[i] = r.GenericParams;
            fors[i] = r.Fors;
            returns[i] = r.Returns;
            scripts[i] = r.Script;
        }

        var genericParamsJson = JsonSerializer.Serialize(genericParams, jsonOptions);
        var forsJson = JsonSerializer.Serialize(fors, jsonOptions);
        var returnsJson = JsonSerializer.Serialize(returns, jsonOptions);

        await using var cmd =
            new NpgsqlCommand(
                """
                insert into actions (id, community_id, annotation, generic_params, fors, returns, script)
                select src.id, src.community, src.annotation, e1.generic_params, e2.fors, e3.returns, src.script
                from unnest($1, $2, $3, $7) with ordinality as src(id, community, annotation, script, idx)
                join lateral jsonb_array_elements($4) with ordinality as e1(generic_params, idx1) on src.idx = idx1
                join lateral jsonb_array_elements($5) with ordinality as e2(fors, idx2) on src.idx = idx2
                join lateral jsonb_array_elements($6) with ordinality as e3(returns, idx3) on src.idx = idx3
                """,
                conn
            );

        // ReSharper disable BitwiseOperatorOnEnumWithoutFlags
        cmd.Parameters.Add(new NpgsqlParameter { Value = ids, NpgsqlDbType = NpgsqlDbType.Array | NpgsqlDbType.Text });
        cmd.Parameters.Add(new NpgsqlParameter
            { Value = communities, NpgsqlDbType = NpgsqlDbType.Array | NpgsqlDbType.Text });
        cmd.Parameters.Add(new NpgsqlParameter
            { Value = annotations, NpgsqlDbType = NpgsqlDbType.Array | NpgsqlDbType.Text });
        // ReSharper restore BitwiseOperatorOnEnumWithoutFlags
        cmd.Parameters.Add(new NpgsqlParameter { Value = genericParamsJson, NpgsqlDbType = NpgsqlDbType.Jsonb });
        cmd.Parameters.Add(new NpgsqlParameter { Value = forsJson, NpgsqlDbType = NpgsqlDbType.Jsonb });
        cmd.Parameters.Add(new NpgsqlParameter { Value = returnsJson, NpgsqlDbType = NpgsqlDbType.Jsonb });
        // ReSharper disable once BitwiseOperatorOnEnumWithoutFlags
        cmd.Parameters.Add(new NpgsqlParameter
            { Value = scripts, NpgsqlDbType = NpgsqlDbType.Array | NpgsqlDbType.Text });

        await cmd.ExecuteNonQueryAsync(cancellationToken);
        return true;
    }

    private static List<T> ReadToList<T>(NpgsqlDataReader reader, Func<NpgsqlDataReader, T> func)
    {
        var result = new List<T>();
        while (reader.Read()) result.Add(func(reader));
        return result;
    }
}

File: src/Rulesage.Common/Repositories/Implementations/ActionRepository.fs
namespace Rulesage.Common.Repositories.Implementations

open System.Text.Json
open Npgsql
open NpgsqlTypes
open Pgvector
open Rulesage.Common.Grammar.Ast
open Rulesage.Common.Repositories.Abstractions

type ActionRepository(dataSource: NpgsqlDataSource, jsonOptions: JsonSerializerOptions) =
    static let readAll (reader: NpgsqlDataReader) (f: NpgsqlDataReader -> 'T) : 'T list =
        [
            while reader.Read() do
                yield f reader
        ]

    let deserialize (json: string) : 'T =
        JsonSerializer.Deserialize<'T>(json, jsonOptions)
    
    let serialize target : string =
        JsonSerializer.Serialize(target, jsonOptions)

    let mapToActionExpr
        (reader: NpgsqlDataReader)
        (idOrd: int)
        (ignoreOrd: int)
        (communityOrd: int)
        (annotationOrd: int)
        (genericParamsOrd: int)
        (forsOrd: int)
        (returnsOrd: int)
        (scriptOrd: int)
        (community: string option)
        =
        let communityValue =
            match community with
            | Some c -> c
            | None -> reader.GetString(communityOrd)

        {
            Id = reader.GetString(idOrd)
            Ignore =
                if community.IsSome then
                    false
                else
                    reader.GetBoolean(ignoreOrd)
            Community = communityValue
            Annotation = reader.GetString(annotationOrd)
            GenericParams = reader.GetString(genericParamsOrd) |> deserialize
            Fors = reader.GetString(forsOrd) |> deserialize
            Returns = reader.GetString(returnsOrd) |> deserialize
            Script = reader.GetString(scriptOrd)
        }

    interface IActionRepository with

        member _.GetDocumentsAsync(cancellationToken) =
            task {
                use conn = dataSource.CreateConnection()
                do! conn.OpenAsync(cancellationToken)
                use cmd = new NpgsqlCommand("SELECT annotation FROM actions", conn)
                use! reader = cmd.ExecuteReaderAsync(cancellationToken)
                let results = readAll reader _.GetString(0)
                return results :> string seq
            }

        member _.FindByIdsAsync(ids, cancellationToken) =
            task {
                use conn = dataSource.CreateConnection()
                do! conn.OpenAsync(cancellationToken)

                use cmd =
                    new NpgsqlCommand(
                        """
                        SELECT
                            id,
                            ignore,
                            community_id,
                            annotation,
                            generic_params,
                            fors,
                            returns,
                            script
                        FROM actions
                        WHERE id = ANY($1)
                        """,
                        conn
                    )

                cmd.Parameters.Add(
                    NpgsqlParameter<string[]>(
                        Value = Seq.toArray ids,
                        NpgsqlDbType = (NpgsqlDbType.Array ||| NpgsqlDbType.Text)
                    )
                )
                |> ignore

                use! reader = cmd.ExecuteReaderAsync(cancellationToken)

                let idOrd = reader.GetOrdinal("id")
                let ignoreOrd = reader.GetOrdinal("ignore")
                let communityOrd = reader.GetOrdinal("community_id")
                let annotationOrd = reader.GetOrdinal("annotation")
                let genericParamsOrd = reader.GetOrdinal("generic_params")
                let forsOrd = reader.GetOrdinal("fors")
                let returnsOrd = reader.GetOrdinal("returns")
                let scriptOrd = reader.GetOrdinal("script")

                let results =
                    readAll
                        reader
                        (fun r ->
                            mapToActionExpr
                                r
                                idOrd
                                ignoreOrd
                                communityOrd
                                annotationOrd
                                genericParamsOrd
                                forsOrd
                                returnsOrd
                                scriptOrd
                                None
                        )

                return results :> ActionExpr seq
            }

        member _.FindOrderByCosineDistanceAsync(contextCommunity, queryVector, skip, take, cancellationToken) =
            task {
                use conn = dataSource.CreateConnection()
                do! conn.OpenAsync(cancellationToken)

                use cmd =
                    new NpgsqlCommand(
                        """
                        SELECT
                            id,
                            annotation,
                            generic_params,
                            fors,
                            returns,
                            script,
                            (annotation_embedding <=> $1) AS distance
                        FROM actions
                        WHERE (community_id = $2 OR community_id = '') AND ignore = false
                        ORDER BY distance
                        LIMIT $3 OFFSET $4
                        """,
                        conn
                    )

                cmd.Parameters.Add(NpgsqlParameter(Value = Vector(queryVector), DataTypeName = "vector"))
                |> ignore

                cmd.Parameters.Add(NpgsqlParameter<string>(Value = contextCommunity, NpgsqlDbType = NpgsqlDbType.Text))
                |> ignore

                cmd.Parameters.Add(NpgsqlParameter<int>(Value = take, NpgsqlDbType = NpgsqlDbType.Integer))
                |> ignore

                cmd.Parameters.Add(NpgsqlParameter<int>(Value = skip, NpgsqlDbType = NpgsqlDbType.Integer))
                |> ignore

                use! reader = cmd.ExecuteReaderAsync(cancellationToken)

                let idOrd = reader.GetOrdinal("id")
                let annotationOrd = reader.GetOrdinal("annotation")
                let genericParamsOrd = reader.GetOrdinal("generic_params")
                let forsOrd = reader.GetOrdinal("fors")
                let returnsOrd = reader.GetOrdinal("returns")
                let scriptOrd = reader.GetOrdinal("script")
                let distanceOrd = reader.GetOrdinal("distance")

                let results =
                    readAll
                        reader
                        (fun r ->
                            let expr =
                                {
                                    Id = r.GetString(idOrd)
                                    Ignore = false
                                    Community = contextCommunity
                                    Annotation = r.GetString(annotationOrd)
                                    GenericParams = r.GetString(genericParamsOrd) |> deserialize
                                    Fors = r.GetString(forsOrd) |> deserialize
                                    Returns = r.GetString(returnsOrd) |> deserialize
                                    Script = r.GetString(scriptOrd)
                                }

                            let similarity = 1.0f - float32 (r.GetDouble(distanceOrd))
                            expr, similarity
                        )

                return results :> seq<ActionExpr * float32>
            }

        member _.SaveAsync(actions, cancellationToken) =
            task {
                let actionsArray = Seq.toArray actions

                let ids = actionsArray |> Array.map _.Id
                let ignores = actionsArray |> Array.map _.Ignore
                let communities = actionsArray |> Array.map _.Community
                let annotations = actionsArray |> Array.map _.Annotation
                let scripts = actionsArray |> Array.map _.Script

                let genericParamsJson =
                    actionsArray |> Array.map _.GenericParams |> serialize

                let forsJson = actionsArray |> Array.map _.Fors |> serialize
                let returnsJson = actionsArray |> Array.map _.Returns |> serialize

                use conn = dataSource.CreateConnection()
                do! conn.OpenAsync(cancellationToken)

                use cmd =
                    new NpgsqlCommand(
                        """
                        INSERT INTO actions (id, ignore, community_id, annotation, generic_params, fors, returns, script)
                        SELECT src.id, src.ignore, src.community, src.annotation, e1.generic_params, e2.fors, e3.returns, src.script
                        FROM unnest($1, $2, $3, $4, $8) WITH ORDINALITY AS src(id, ignore, community, annotation, script, idx)
                        JOIN LATERAL jsonb_array_elements($5) WITH ORDINALITY AS e1(generic_params, idx1) ON src.idx = idx1
                        JOIN LATERAL jsonb_array_elements($6) WITH ORDINALITY AS e2(fors, idx2) ON src.idx = idx2
                        JOIN LATERAL jsonb_array_elements($7) WITH ORDINALITY AS e3(returns, idx3) ON src.idx = idx3
                        """,
                        conn
                    )

                cmd.Parameters.Add(
                    NpgsqlParameter(Value = ids, NpgsqlDbType = (NpgsqlDbType.Array ||| NpgsqlDbType.Text))
                )
                |> ignore

                cmd.Parameters.Add(
                    NpgsqlParameter(Value = ignores, NpgsqlDbType = (NpgsqlDbType.Array ||| NpgsqlDbType.Boolean))
                )
                |> ignore

                cmd.Parameters.Add(
                    NpgsqlParameter(Value = communities, NpgsqlDbType = (NpgsqlDbType.Array ||| NpgsqlDbType.Text))
                )
                |> ignore

                cmd.Parameters.Add(
                    NpgsqlParameter(Value = annotations, NpgsqlDbType = (NpgsqlDbType.Array ||| NpgsqlDbType.Text))
                )
                |> ignore

                cmd.Parameters.Add(NpgsqlParameter(Value = genericParamsJson, NpgsqlDbType = NpgsqlDbType.Jsonb))
                |> ignore

                cmd.Parameters.Add(NpgsqlParameter(Value = forsJson, NpgsqlDbType = NpgsqlDbType.Jsonb))
                |> ignore

                cmd.Parameters.Add(NpgsqlParameter(Value = returnsJson, NpgsqlDbType = NpgsqlDbType.Jsonb))
                |> ignore

                cmd.Parameters.Add(
                    NpgsqlParameter(Value = scripts, NpgsqlDbType = (NpgsqlDbType.Array ||| NpgsqlDbType.Text))
                )
                |> ignore

                return! cmd.ExecuteNonQueryAsync(cancellationToken)
            }
